### DISCUSSION TITLE
How to see only differences

### DISCUSSION THREAD

**Discussion by eddydm57** [ImageMagick/ImageMagick#7300] [Digital Image Processing]
I'm trying to draw with graphite pencils, so only black color is used. At this moment I have the habbit to scan the result of my drawing in different stages.
I'v already used the imagick program to align the different scans and to process different images, but i don't succeed to have only the differences. I always get also the common parts.

Let me explain more in detail. (Mostly I work in powershell)
Lets say i'v 2 stages saved in files $f1 and $f2
I'm using the next instructions
        $f4 = "test4.jpg"
        magick.exe $f1 -fill yellow +opaque white $f4
        $f5 = "test5.jpg"
        magick.exe $f2 -fill red +opaque white $f5
        magick.exe composite $f4 $f5 -compose difference  difference_jpeg.gif
I hoped the drawing in the first stage was converted from black to yellow,
I hoped the drawing in the second stage was converted to a red drawing
I hoped that by composing the difference a would get a third drawing with
* yellow areas for lines in the first stage that were lost in second stage (so corrections)
* red areas for the new lines in the second stage
* combination of red and yellow (=orange) for the areas that weren't changed. The color isn't orange but i see at as a kind of green??? How can i change that color to white.

Thanks in advance for any idea's.

Eddy

**Comment by snibgo** [comment]
Suppose your images have three channels: red, green and blue. On a scale of 0 to 255, a "red" pixel has channel values (255,0,0), and "yellow" pixels are (255,255,0).

A "difference" composite is the absolute values in each channel of one pixel minus the other.

So the "difference" between red and yellow is: (255-255, 255-0, 0-0) = (0,255,0), which is green.

Note that the difference between two identical colour is black.

In your input images, suppose you have only black and white pixels. So we have four combinations of inputs, each with a desired output.
```
img1   img2   output
===    ====   ======
black  black  orange
black  white  yellow
white  black  red
white  white  orange
```
There are many ways to do this. One is to make the new image where the red channel comes from img1, and the green and blue channels from img2. This will give:
```
img1   img2   output
===    ====   ======
black  black  black
black  white  cyan
white  black  red
white  white  white
```
Then we can change black to orange with "`-fill orange -opaque black`", and cyan to yellow, and white to orange.

Windows BAT command (untested):
```
magick ^
  img1.png img2.png +duplicate ^
  -combine ^
  -fill orange -opaque black ^
  -fill yellow -opaque cyan ^
  -fill orange -opaque white ^
  out.png
```
If you need further help, please post sample inputs.

**Reply by eddydm57** [reply]
Thanks a lot for this idea.
I have a lot to study about the duplicate action, and combine and about the colour scheme.
But at this moment I accept your theory and it seems to work! Many thanks!
